### PR TITLE
Bidirectional Linking

### PR DESCRIPTION
**Pull request by jvorob**
Adds bidirectional linking (same-table cursor-links)

Allows linking sections in a cycle as long as all are same-table cursor links
(e.g. A->B->A,  A->B->C->A, etc)

Clicking/moving cursor in any of the linked sections updates them all.
Works correctly even if sections have differing filters

Tests now check that same-table cursor cycles are allowed in selectBy,
but others are forbidden.

No actual tests of behavior, though

### CODE REVIEW

File: app/client/ui/selectBy.ts [hunk 1]
```diff
@@ -49,8 +49,13 @@ interface LinkNode {
...
 
   // list of ids of the sections that are ancestors to this section according to the linked section
-  // relationship
-  ancestors: Set<number>;
+  // relationship. ancestors[0] is this.section, ancestors[length-1] is oldest ancestor
+  ancestors: number[];
+
+  // Records which links between ancestors are same-table cursor-links
+  // if isAncCursLink[0] == true, that means the link from ancestors[0] to ancestors[1] is a same-table cursor-link
+  // NOTE: (it is therefore 1 shorter than ancestors, unless the links form a cycle in which case it's the same len)
```
Comment (berhalak) [inline app/client/ui/selectBy.ts:57]: Can you expand here - I don't understand it .

File: app/client/ui/selectBy.ts [hunk 2]
```diff
@@ -141,9 +146,35 @@ function isValidLink(source: LinkNode, target: LinkNode) {
...
   }
 
-  // The link must not create a cycle
-  if (source.ancestors.has(target.section.getRowId())) {
-    return false;
+  // The link must not create a cycle, unless it's only same-table cursor-links all the way to target
+  if (source.ancestors.includes(target.section.getRowId())) {
+
+    //cycles only allowed for cursor links
+    if(source.column || target.column || source.isSummary) {
```
Comment (berhalak) [inline app/client/ui/selectBy.ts:153]:
```suggestion
    if (source.column || target.column || source.isSummary) {
```

File: app/client/ui/selectBy.ts [hunk 3]
```diff
@@ -141,9 +146,35 @@ function isValidLink(source: LinkNode, target: LinkNode) {
     }
   }
 
-  // The link must not create a cycle
-  if (source.ancestors.has(target.section.getRowId())) {
-    return false;
+  // The link must not create a cycle, unless it's only same-table cursor-links all the way to target
+  if (source.ancestors.includes(target.section.getRowId())) {
+
+    //cycles only allowed for cursor links
```
Comment (berhalak) [inline app/client/ui/selectBy.ts:152]:
```suggestion
    // cycles only allowed for cursor links
```

File: app/client/ui/selectBy.ts [hunk 4]
```diff
@@ -141,9 +146,35 @@ function isValidLink(source: LinkNode, target: LinkNode) {
...
+        // We made it! All is well
+        break;
+      }
+      // If we've hit the last ancestor and haven't found target, error out (shouldn't happen, we checked for it)
+      if(i == source.ancestors.length-1) { throw Error("Error: Array doesn't include targetSection"); }
+
+      // Need to follow another link back, verify that it's a same-table cursor-link
+      if(!source.isAncestorSameTableCursorLink[i]) { return false; }
+    }
+    console.log("===== selectBy found valid cycle", JSON.stringify(source)); //TODO JV TEMP DEBUG
```
Comment (berhalak) [inline app/client/ui/selectBy.ts:176]:
```suggestion

```

File: app/client/ui/selectBy.ts [hunk 5]
```diff
@@ -141,9 +146,35 @@ function isValidLink(source: LinkNode, target: LinkNode) {
...
+
+    // Walk backwards along the chain of ancestors
+    // - if we hit a non-cursor link before reaching target, then that would be an illegal cycle
+    // - when we hit target, we've verified that this is a legal cycle, so break
+    //   (ancestors further up the hierarchy past target don't matter, since once we set target.linkSrcSec = src.sec,
+    //    they would stop being ancestors of src)
+    // NOTE: we're guaranteed to hit target before the end of the array (because of the `if(...includes...)` above)
+    // ALSO NOTE: isAncestorSameTableCursorLink may be 1 shorter than ancestors, but it's accounted for by the above
+
+    for(let i = 0; ; i++) {
```
Comment (berhalak) [inline app/client/ui/selectBy.ts:165]: Can you use here the while loop. I haven't notice the lack of condition for a while.

File: app/client/ui/selectBy.ts [hunk 6]
```diff
@@ -141,9 +146,35 @@ function isValidLink(source: LinkNode, target: LinkNode) {
...
+      if(source.ancestors[i] == target.section.getRowId()) {
+        // We made it! All is well
+        break;
+      }
+      // If we've hit the last ancestor and haven't found target, error out (shouldn't happen, we checked for it)
+      if(i == source.ancestors.length-1) { throw Error("Error: Array doesn't include targetSection"); }
```
Comment (berhalak) [inline app/client/ui/selectBy.ts:171]:
```suggestion
      if (i == source.ancestors.length-1) { throw Error("Error: Array doesn't include targetSection"); }
```

File: app/client/ui/selectBy.ts [hunk 7]
```diff
@@ -141,9 +146,35 @@ function isValidLink(source: LinkNode, target: LinkNode) {
...
+
+      // Need to follow another link back, verify that it's a same-table cursor-link
+      if(!source.isAncestorSameTableCursorLink[i]) { return false; }
```
Comment (berhalak) [inline app/client/ui/selectBy.ts:174]:
```suggestion
      if (!source.isAncestorSameTableCursorLink[i]) { return false; }
```

File: app/client/ui/selectBy.ts (hunk 6 above)
Comment (berhalak) [inline app/client/ui/selectBy.ts:171]:
```suggestion
      if(i == source.ancestors.length-1) { throw Error("Array doesn't include targetSection"); }
```

File: app/client/ui/selectBy.ts [hunk 8]
```diff
@@ -141,9 +146,35 @@ function isValidLink(source: LinkNode, target: LinkNode) {
...
+
+    for(let i = 0; ; i++) {
+      if(source.ancestors[i] == target.section.getRowId()) {
+        // We made it! All is well
+        break;
+      }
+      // If we've hit the last ancestor and haven't found target, error out (shouldn't happen, we checked for it)
+      if(i == source.ancestors.length-1) { throw Error("Error: Array doesn't include targetSection"); }
+
+      // Need to follow another link back, verify that it's a same-table cursor-link
```
Comment (berhalak) [inline app/client/ui/selectBy.ts:173]: Can you rephrase it?

File: app/client/ui/selectBy.ts [hunk 9]
```diff
@@ -224,15 +255,29 @@ function fromViewSectionRec(section: ViewSectionRec): LinkNode[] {
...
+      // There's a cycle in the existing link graph
+      // TODO if we're feeling fancy, can test here whether it's an all-same-table cycle and warn if not
+      //      but there's other places we check for that
       break;
     }
-    ancestors.add(sec.getRowId());
+    ancestors.push(sec.getRowId());
+
+    //Determine if this link is a same-table cursor link
+    if(sec.linkSrcSection.peek().getRowId()) {
```
Comment (berhalak) [inline app/client/ui/selectBy.ts:272]:
```suggestion
    if (sec.linkSrcSection.peek().getRowId()) {
```

Why do we need this condition? Isn't it already checked in the for loop?

File: app/client/ui/selectBy.ts [hunk 10]
```diff
@@ -224,15 +255,29 @@ function fromViewSectionRec(section: ViewSectionRec): LinkNode[] {
...
+      const srcCol = sec.linkSrcCol.peek().getRowId();
+      const tgtCol = sec.linkTargetCol.peek().getRowId();
+      const srcTable = sec.linkSrcSection.peek().table.peek();
+      const srcIsSummary = srcTable.primaryTableId.peek() !== srcTable.tableId.peek();
+      isAncestorSameTableCursorLink.push(srcCol == 0 && tgtCol == 0 && !srcIsSummary);
```
Comment (berhalak) [inline app/client/ui/selectBy.ts:277]:
```suggestion
      isAncestorSameTableCursorLink.push(srcCol === 0 && tgtCol === 0 && !srcIsSummary);
```

File: app/client/ui/selectBy.ts [hunk 11]
```diff
@@ -224,15 +255,29 @@ function fromViewSectionRec(section: ViewSectionRec): LinkNode[] {
...
+    }
+    // NOTE: isAncestorSameTableCursorLink may be 1 shorter than ancestors, since last ancestor has no incoming link
```
Comment (berhalak) [inline app/client/ui/selectBy.ts:279]: Can you explain more here? From reading the code it looks opposite.

File: test/nbrowser/RightPanelSelectBy.ts [hunk 12]
```diff
@@ -63,9 +63,23 @@ describe('RightPanelSelectBy', function() {
...
+
+    // this link should no longer be present, since it would create a cycle with a filter link in it
     await gu.openSelectByForSection('PERFORMANCES RECORD');
-    assert.equal(await driver.findContent('.test-select-row', /Performances detail/).isPresent(), false);
+    assert.equal(await driver.findContent('.test-select-row', /Performances record.*Film/i).isPresent(), false);
+  });
+
+  it('should allow creating cursor-linked-cycles', async function() {
+    assert.equal(await driver.findContent('.test-select-row', /Performances detail/).isPresent(), true);
+
```
Comment (berhalak) [inline test/nbrowser/RightPanelSelectBy.ts:80]: Are you planning to add test, or will it be hidden for a while?

File: test/nbrowser/RightPanelSelectBy.ts [hunk 13]
```diff
@@ -63,9 +63,23 @@ describe('RightPanelSelectBy', function() {
...
+    // undo, link is expected to be unset for next test
```
Comment (berhalak) [inline test/nbrowser/RightPanelSelectBy.ts:81]:
```suggestion
    // undo the operation in the previous test,, link is expected to be unset for next test
```

File: app/client/components/GristDoc.ts [hunk 14]
```diff
@@ -1132,6 +1133,12 @@ export class GristDoc extends DisposableWithEvents {
       if (!section.id.peek()) {
         throw new Error(`Section ${cursorPos.sectionId} does not exist`);
       }
+
+      if (visitedSections.includes(section.id.peek())) {
+        //We've already been here (we hit a cycle), just return immediately
```
Comment (berhalak) [inline app/client/components/GristDoc.ts:1138]:
```suggestion
        // We've already been here (we hit a cycle), just return immediately
```

File: app/client/components/LinkingState.ts [hunk 15]
```diff
@@ -17,6 +17,7 @@ import merge = require('lodash/merge');
 import mapValues = require('lodash/mapValues');
 import pick = require('lodash/pick');
 import pickBy = require('lodash/pickBy');
+import {SequenceNEVER, SequenceNum} from "./Cursor";
```
Comment (berhalak) [inline app/client/components/LinkingState.ts:20]: Sort imports and absolute paths

File: app/client/components/LinkingState.ts [hunk 16]
```diff
@@ -194,13 +200,111 @@ export class LinkingState extends Disposable {
...
+      // If multiple viewSections are cursor-linked together A->B->C, we need to propagate the linked cursorPos along.
+      // The old way was to have: A.activeRowId -> (sets by cursor-link) -> B.activeRowId, and so on
+      //                                                                                                               |
+      //                                   -->  [B.LS]                    --> [C.LS]                                   |
+      //                                  /        | B.LS.cursorPos      /       | C.LS.cursorPos                      |
+      //                                 /         v                    /        v                                     |
+      //                   [ A ]--------/        [ B ]   --------------/       [ C ]                                   |
+      //                        A.actRowId                B.actRowId                                                   |
+      //
+      // However, if e.g. viewSec B is filtered, the correct rowId might not exist in B, and so it's rowId would be
```
Comment (berhalak) [inline app/client/components/LinkingState.ts:214]:
```suggestion
      // However, if e.g. viewSec B is filtered, the correct rowId might not exist in B, and so it's activeRowId would be
```

File: app/client/components/LinkingState.ts [hunk 17]
```diff
@@ -194,13 +200,111 @@ export class LinkingState extends Disposable {
...
+      //                                /        |                          |                                          |
+      //                               /         v B.LS.cursorPos           v C.LS.cursorPos                           |
+      //                 [ A ]--------/        [ B ]                      [ C ]                                        |
+      //                      A.actRowId                                                                               |
+      //
+      // If the previous section has a linkingState, we use the previous LS's incomingCursorPos
+      // (i.e. two sections back) instead of looking at our srcSection's activeRowId. This way it doesn't matter how
+      // section B is filtered, since we're getting our cursorPos straight from A (through a computed in B.LS)
+      //
+      // However, each linkingState needs to decide whether to use the cursorPos from the srcSec (i.e. its activeRowId),
```
Comment (berhalak) [inline app/client/components/LinkingState.ts:236]: Why it neeeds to decide it?

File: app/client/components/LinkingState.ts [hunk 18]
```diff
@@ -194,13 +200,111 @@ export class LinkingState extends Disposable {
...
+      // or to use the previous linkState's incomingCursorPos. We want to use whichever section the user most recently
+      // interacted with, i.e. whichever cursor update was most recent. For this we use, the cursor version (given in
+      // viewSection.lastCursorEdit). incomingCursorPos is a pair of [rowId, sequenceNum], so each linkingState sets its
+      // incomingCursorPos to whichever is most recent between its srcSection, and the previous LS's incCursPos.
+      //
+      // The end result is that there is that because the lastCursorEdits are guaranteed to be unique, there is always
```
Comment (berhalak) [inline app/client/components/LinkingState.ts:242]:
```suggestion
      // The end result is that because the lastCursorEdits are guaranteed to be unique, there is always
```
But not sure what this means

File: app/client/components/LinkingState.ts [hunk 19]
```diff
@@ -194,13 +200,111 @@ export class LinkingState extends Disposable {
...
-        this.cursorPos = this.autoDispose(ko.computed(() =>
-          srcValueFunc(srcSection.activeRowId()) as UIRowId
-        ));
+      // check for failure
+      if (srcValueFunc) {
+        //Incoming-cursor-pos determines what the linked cursor position should be, considering the previous
+        //linked section (srcSection) and all upstream sections (through srcSection.linkingState)
+        this.incomingCursorPos = this.autoDispose((ko.computed(() => {
+          // Get previous linkingstate's info
+          // NOTE: prevlink = this.srcSec.linkingState is 2 hops back, (i.e. it reads from from this.srcSec.linkSrcSec)
```
Comment (berhalak) [inline app/client/components/LinkingState.ts:268]:
```suggestion
          // NOTE: prevlink = this.srcSec.linkingState is 2 hops back, (i.e. it reads from this.srcSec.linkSrcSec)
```

Can you rewrite it and add more explanation I don't follow this..

File: app/client/components/LinkingState.ts [hunk 20]
```diff
@@ -194,13 +200,111 @@ export class LinkingState extends Disposable {
...
+            (prevLink.linkTypeDescription() == "Cursor:Same-Table" ||
+             prevLink.linkTypeDescription() == "Cursor:Reference");
+          const [prevLinkedPos, prevLinkedVersion] = prevLinkHasCursor ? prevLink.incomingCursorPos() :
+                                                                         [null, SequenceNEVER];
+          // Get srcSection's info
+          const srcSecPos = this._srcSection.activeRowId.peek(); //we don't depend on this, only on its cursor version
+          const srcSecVersion = this._srcSection.lastCursorEdit();
+
+          // Sanity check: version might be NEVER if viewSection's cursor hasn't initialized (shouldn't happen I think)
+          if(srcSecVersion == SequenceNEVER) {
```
Comment (berhalak) [inline app/client/components/LinkingState.ts:281]:
```suggestion
          if(srcSecVersion === SequenceNEVER) {
```

**Review by berhalak — CHANGES_REQUESTED** [review]

**Comment by berhalak** [comment]
Have you thought about the UI changes it should bring along?

File: app/client/components/Cursor.ts [hunk 21]
```diff
@@ -93,8 +117,18 @@ export class Cursor extends Disposable {
...
+
+    // update the section's activeRowId and lastCursorEdit when needed
     this.autoDispose(this._properRowId.subscribe((rowId) => baseView.viewSection.activeRowId(rowId)));
+    this.autoDispose(this._lastEditedAt.subscribe((seqNum) => baseView.viewSection.lastCursorEdit(seqNum)));
+
+    // Update the cursor edit time if either the row or column change
+    // IMPORTANT: need to subscribe AFTER the properRowId->activeRowId subscription.
+    //  (Cursor-linking observables depend on lastCursorEdit, but only peek at activeRowId. Therefore, updating the
+    //   edit time triggers a re-read of activeRowId, and swapping the order will read stale values for rowId)
+    this.autoDispose(this._properRowId.subscribe(() => { this.cursorEdited(); }));
```
[the remark this answers is not in the document]
Comment (jvorob) [reply app/client/components/Cursor.ts:130]: don't think so, added comment

File: app/client/components/Cursor.ts [hunk 22]
```diff
@@ -116,9 +150,17 @@ export class Cursor extends Disposable {
...
    * preferring rowId.
+   *
+   * isFromLink prevents lastEditedAt from being updated, so lastEdit reflects only user-driven edits
    * @param cursorPos: Position as { rowId?, rowIndex?, fieldIndex? }, as from getCursorPos().
+   * @param isFromLink: should be set if this is a cascading update from cursor-linking
    */
-  public setCursorPos(cursorPos: CursorPos): void {
+  public setCursorPos(cursorPos: CursorPos, isFromLink: boolean = false): void {
+    //If updating as a result of links, we want to NOT update lastEditedAt
+    if(isFromLink) { this._silentUpdatesFlag = true; }
```
[the remark this answers is not in the document]
Comment (jvorob) [reply app/client/components/Cursor.ts:160]: ooh, good catch

File: app/client/components/LinkingState.ts (hunk 17 above)
Comment (jvorob) [reply app/client/components/LinkingState.ts:236]: As in:

the linkingState can take cursorPos from this.srcSection or this.srcSection.linkingState

If the srcSection is "driving" the linking, i.e. was most recently clicked, then we use it (e.g. B.LS pulls from its srcSection directly). 
Otherwise the driver is further up the chain, so we use srcSection.linkingState to get to it (e.g. C.LS pulls from its srcSection.linkingState, since B isn't the driver of this chain)

Whichever it is, there needs to be some logic that decides which to use, is what I was trying to say.

Does that explain it better? And/or is the original explanation misleading?

File: app/client/ui/selectBy.ts (hunk 9 above)
Comment (jvorob) [reply app/client/ui/selectBy.ts:272]: We're one step ahead of the loop, the loop test for sec.linkSrcSection won't run until we're finished with this iteration, and I believe we end up going 1 past the end of the loop anyway, so we'll have to make some extra checks

File: app/client/ui/selectBy.ts (hunk 11 above)
Comment (jvorob) [reply app/client/ui/selectBy.ts:279]: If we have A->B->C, the loop will go C, B, and stop at A, because A has previous node
However, the link-check will only do C and B, and skip A, since it has no incoming link

File: test/nbrowser/RightPanelSelectBy.ts (hunk 12 above)
Comment (jvorob) [reply test/nbrowser/RightPanelSelectBy.ts:80]: I don't understand the question. Do you mean like, am I planning to extend the "should allow cycles" test?

I just split the test into two parts, verifying that it forbids invalid cycles, and then verify that it allows cursor cycles, but the second part ended up being a oneliner

**Comment by jvorob** [comment]
RE: Have I thought about UI changes it should bring

- SelectBy needs to show cycles so you can select them in the first place (already done as part of this PR)

- link info (in rightPanel) could maybe better indicate that things are cyclically linked instead of the current thing which just says "Link from section A sets cursor to [blah]"? Maybe could indicate which section was last clicked in / is driving the link? But that's non-trivial / might require walking the graph, and it's unclear if that's even a useful feature, so I don't think it needs to be in this diff

- Some way to show when rows are filtered out, especially now that cursor-linking can leapfrog a section, which might be very confusing to users. However, there's nowhere in the UI to pus this at the moment. I think it would go well with the link-state bubbles I was working on earlier this summer, but they got sidelined since we couldn't get consensus on UI. However, I'm going to clean up and re-post that diff before I'm done here, so hopefully the code will all be there once people decide where to display the UI, and that would be a very natural place to indicate broken cursor-linking due to filtering (e.g. the linkstate bubble goes red, on hover it explains that "I'm trying to show this row because of cursor linking, but it's filtered out here")

**Comment by jperon** [comment]
@jvorob Thank you very much for this work, which would be really handy!
With such a feature, it would be nice to enable the "select widget" on cards that are linked that way ; otherwise, they can’t really be used as selector, as can be viewed on [this sample](https://grist-link-bidirectional.fly.dev/iwakvvj9PiNE/Card-widget#a1.s3.r1.c2):

CardSelectWidget

**Review by berhalak — CHANGES_REQUESTED** [review]

File: app/client/ui/selectBy.ts [hunk 23]
```diff
@@ -141,9 +148,36 @@ function isValidLink(source: LinkNode, target: LinkNode) {
...
+
+    // Walk backwards along the chain of ancestors
+    // - if we hit a non-cursor link before reaching target, then that would be an illegal cycle
+    // - when we hit target, we've verified that this is a legal cycle, so break
+    //   (ancestors further up the hierarchy past target don't matter, since once we set target.linkSrcSec = src.sec,
+    //    they would stop being ancestors of src)
+    // NOTE: we're guaranteed to hit target before the end of the array (because of the `if(...includes...)` above)
+    // ALSO NOTE: isAncestorSameTableCursorLink may be 1 shorter than ancestors, but it's accounted for by the above
+    let i = 0;
+    for (;;) {
```
[the remark this answers is not in the document]
Comment (jvorob) [reply app/client/ui/selectBy.ts:167]: So, there's already a `if i == source.ancestors.length - 1 { throw Error }` in the body of the loop to keep it from going infinite. Maybe I could change it to `if i >=` to make it a little safer? But I think trying to push that condition into the while would be ugly, since the success condition is breaking out of the loop. Would have to do some sort of thing with a `successFlag = false`,   `if(successCondition) { successFlag = true; break; } ... /* after loop */ if(!successFlag) {  throw Error }`, whereas  the current code already covers it. Do you think it's acceptable to leave it as is or should I force that condition into the while?

**Comment by berhalak** [comment]
> @jvorob Thank you very much for this work, which would be really handy! With such a feature, it would be nice to enable the "select widget" on cards that are linked that way ; otherwise, they can’t really be used as selector, as can be viewed on [this sample](https://grist-link-bidirectional.fly.dev/iwakvvj9PiNE/Card-widget#a1.s3.r1.c2):

Hi @jperon,

Can you explain what you have in mind by "on cards". The "Select By" is enabled for all kind of widgets by default, so a "Card list" widget is supported out of box. A single "Card Widget" is also supported, but as it doesn't have navigation buttons, this feature wouldn't make much sense there.

**Comment by dsagal** [comment]
> Can you explain what you have in mind by "on cards". The "Select By" is enabled for all kind of widgets by default, so a "Card list" widget is supported out of box. A single "Card Widget" is also supported, but as it doesn't have navigation buttons, this feature wouldn't make much sense there.

A Card widget sometimes has navigation buttons, so maybe @jperon is suggesting to enable those in this case?

image

**Comment by jperon** [comment]
@dsagal You have exactly seen what I meant.

**Review by berhalak — APPROVED** [review]
Thank you @jvorob !
